server/controllers/auth/signup.js

```diff
@@ -1,3 +1,34 @@
-exports.signup = (req, res) => {
-  res.send(' Hello from signup file');
+const { hash } = require('bcrypt');
+const { jwtSign } = require('../../helpers');
+const { signupSchema } = require('./validation/signup-validation');
+const { users: { getUserByEmailOrUsername, addUser } } = require('../../database/queries');
+
+
+exports.signup = (req, res, next) => {
+  const key = process.env.KEY;
+  const {
+    username, email, password, nativeLangId, learningLangId,
+  } = req.body;
+
+  signupSchema.validate({
+    username, email, password, nativeLangId, learningLangId,
+  })
+    .then(() => getUserByEmailOrUsername(email, username))
+    .then(({ rows }) => {
+      if (rows.length !== 0) throw Error('username or email exists');
```

Comment:
I think it won't be a good experience for the user to get this message: `username or email exists` since he/she won't be able to know exactly which field is the already taken one, the email or the username :laughing: 

Keep this in mind next time!

Reply:
exactly as @IsraaSulaiman  say 
 - so please separate the query and send specific msg to the user 
   
when you are developing your code keep in your mind the user, the user should know what is the specific error.
 However, we took time when writing the issues not just to make relates. but to be a reference point for us

Reply:
good point! :) but it should be easily fixable:
```
if (rows.length >= 0) {
    if (rows[0].username === username) throw new Error('username exists')
    else if (rows[0].email === email) throw new Error('email exists')
}
```